source/language/types.rst

```diff
@@ -373,3 +375,152 @@ variables whose values may only be known at run time.
+has more than 1 bit of precision.
+
+Casting from int/uint
+~~~~~~~~~~~~~~~~~~~~~
+
+``int[n]`` and ``uint[n]`` values cast to the standard types mimicking ``C``
+behavior. Casting to ``bit[m]`` is only allowed when ``m==n``. If the target 
+``bit`` has more or less precision, then explicit slicing syntax must be given.
+
+Casting from float
+~~~~~~~~~~~~~~~~~~
+
+``float[n]`` values cast to the standard types mimicking ``C`` behavior. 
+Casting a ``float`` value to an ``angle[m]`` is accomplished by first 
+performing a modulo 2π operation on the float value. The resulting value 
+is then converted to the nearest ``angle[m]`` possible. In the event of a 
+tie between two possible nearest values the result is the one with an even 
+least significant bit (*i.e* round to nearest, ties to even).
+
+Casting from angle
+~~~~~~~~~~~~~~~~~~
+
+``angle[n]`` values cast to ``bool`` using the convention ``val != 0.0``. 
+Casting to ``float[m]`` is accomplished by first converting the angle value 
+to the nearest float, with ties going to the value with the even least 
+significant bit, and then multiplying the resulting value by 2π as a 
+``float[m]`` value. Casting to ``bit[m]`` values is only allowed when 
+``n==m``, otherwise explicit slicing synatx must be provided. 
+
+When casting between angles of differing precisions (``n!=m``): if the target 
+has more significant bits, then the value is padded with ``m-n`` least 
+significant bits of ``0``; if the target has fewer significant bits, then 
+there are two acceptable behaviors that can be supported by compilers: 
+rounding and truncation. For rounding the value is rounded to the nearest 
+value, with ties going to the value with the even least significant bit.
+Trunction is likely to have more hardware support. This behavior can be
+controlled by the use of a ``#pragma``.
+
+Casting from bit
+~~~~~~~~~~~~~~~~
+
+``bit[n]`` values cast to ``bool`` using the convention ``val != 0``. Casting to 
+``int[m]`` or ``uint[m]`` is done assuming a little endian 2's complement 
+representation, and is only allowed when ``n==m``, otherwise explicit slicing 
+syntax must be given. Likewise, ``bit[n]`` can only be cast to ``angle[m]`` 
+when ``n==m``, in which case an exact per-bit copy is done using little-endian 
+bit order. Finally, casting between bits of differing precisions is not 
+allowed, explicit slicing syntax must be given.
+
+.. _divideDuration:
+
+Converting duration to other types
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Casting from or to duration values is not allowed, however, operations on 
+durations that produce values of different types is allowed. For example, 
+dividing a duration by a compile-time constant produces a machine-precision 
+``float``.
+
+.. code-block:: c
+
+   const one_ns = 1e-9;
+   duration a = 500ns;
+   float a_in_ns = a / one_ns; // results in 500.0
+
+   float one = 1.0;
+   duration b = 1ms;
+   float b_float = b / one; // results in 1e-3
```

Comment:
These examples seem wrong to me - I expected the first to result in `5e11ns` (or `500s`), and the second to result in `1ms`.  I thought what we'd said was just defining `duration / duration -> float`, with the idea that that avoided specifying a "natural" time unit (e.g. these ones treat the second as the natural unit, since `1ms / 1.0` goes to `1e-3`).

So I was expecting to have

```openqasm
duration one_ns = 1ns;
duration a = 500ns;
float a_in_ns = a / one_ns;  // 500.0

duration one_s = 1s;
float a_in_s = a / one_s; // 5e-7
```

and then `duration / float -> duration`, like "normal" division.

Reply:
This is a good point. I remember that part of the conversation now that you point it out. I've corrected the example to follow your suggestion above in https://github.com/Qiskit/openqasm/pull/268/commits/93e759e1cb14a14c5de6fdae55ffd33ae963ecbf